OpenRA.Mods.Common/Traits/SupportPowers/ProduceActorPower.cs

```diff
@@ -57,15 +57,21 @@ public override void Activate(Actor self, Order order, SupportPowerManager manag
 			base.Activate(self, order, manager);
 
 			var info = Info as ProduceActorPowerInfo;
-			var sp = self.TraitsImplementing<Production>()
-				.FirstOrDefault(p => p.Info.Produces.Contains(info.Type));
+			var producers = self.World.ActorsWithTrait<Production>()
+				.Where(x => x.Actor.Owner == self.Owner
+					&& !x.Trait.IsTraitDisabled
+					&& x.Trait.Info.Produces.Contains(info.Type))
+					.OrderByDescending(x => x.Actor.IsPrimaryBuilding())
+					.ThenByDescending(x => x.Actor.ActorID);
 
 			// TODO: The power should not reset if the production fails.
 			// Fixing this will require a larger rework of the support power code
 			var activated = false;
 
-			if (sp != null)
+			if (producers.Any())
 			{
+				var p = producers.First();
```

Comment:
There is no guarantee that `p.Trait.Produce` will succeed (e.g. the building may be blocked).  It would be better to loop over `producers` until `activated` is true.